# Tutorial_1 Das BWKI Hundetutorial  🐶🩺 

## Musterlösung für Kapitel 1 & 2  📚

Dies ist die Musterlösung zu den Programmieraufgaben des ersten Kapitels.
Schaut sie euch erst an wenn ihr versucht habt, die Übung selbstständig zu lösen und nicht mehr weiter kommt. Bei Fragen könnt ihr euch jederzeit auf unserem Discord Server mit anderen TeilnehmerInnen austauschen!

Viel Erfolg bei diesem Tutorial!

Als erstes müssen wir unsere **Daten einlesen** und für die Datenverarbeitung benötigte Packages(Bibliotheken) einlesen.


In [4]:
#Installiere numpy, ein Paket zur Datenverarbeitung
import numpy as np

#Beschreibung von Paketen erhalten

In [5]:
#Prüfe in welchem Arbeitsverzeichnis du gerade arbeitest (working directory)
#Im selben Ordner müssen auch die Daten gespeichert sein
%pwd

'/home/jovyan/ML/Erkranktet Hunde'

In [6]:
#Du bist noch nicht im richtigen Ordner? Ändere mit cd /Pfad deine working directory dorthin wo die Daten liegen
# ersetze die ... falls du nicht weiter kommst, einfach im Internet nach change working directory Jupyter Notebook suchen
cd /Users/...

NameError: name 'cd' is not defined

In [7]:
#Einlesen der Daten
D = np.load('train_data.npy')

In [8]:
# Eigenschaften des Datensatzes einsehen (Zeilen, Spalten)
D.shape

(426, 14)

Zuerst wollen wir uns einen **Überblick über den Datensatz** verschaffen und uns die 14 verschiedenen Variablen anschauen, die uns für insgesamt 426 Hunde vorliegen. Wichtig: Die Überschriften der Merkmale sind nicht im Datensatz gespeichert, daher sollten wir uns diese am besten abspeichern (z.B. in unserem Jupyter Notebook). Wir müssen mehrfach auf verschiedene Merkmale zurück greifen.

Der Datensatz umfasst folgende 14 Eigenschaften [Python zählt von 0 aufwärts]:

Spalte/ Merkmal 
<br> 0 Alter
<br>1 Geschlecht (0 = männlich, 1 = weiblich)
<br>2 Würmer (0 = keine, 1 = Typ I, 2 = Typ II, 3 = Typ I + II)
<br>3 Ruhender Blutdruck
<br>4 Cholesterinspiegel (in mg/dl)
<br>5 Nüchtern-Blutzuckerspiegel > 120 mg/dl (1 = hoch; 0 = niedrig)
<br>6 Ruhende elektrokardiographische Ergebnisse (0 = normal, 1 = ST-T Abnormalität, 2 = linke ventrikuläre Hypertrophie)
<br>7 Maximale Herzfrequenz
<br>8 Durch Bewegung verursachte Angina pectoris (1 = Ja; 0 = Nein)
<br>9 ST-Absenkung, die durch Bewegung in Bezug auf die Ruhephase hervorgerufen wird (ein Marker im EKG).
<br>10 Die maximale Steigung des ST-Segments bei Bewegung
<br>11 Anzahl der großen Gefäße (0-3), aufgenommen durch Ultraschall
<br>12 Blutfluss im Herz (3 = normal; 6 = fixer defekt; 7 = reversibler defekt )
<br>13 Diagnose (1 = krank, 0 = gesund)

## Kapitel 2

Deine Aufgabe ist es die Herzfrequenz(8. Merkmal) des 10. Hundes herausfinden (Abschnitt 5/20).
Dies ist die MUSTERLÖSUNG! Die Coding-Konsole im Ki-Kurs enthält weitere Informationen zu Funktionen, Argumenten oder Erklärungen zum Ergebnis.

In [ ]:
herzfrequenz = D[9,7]
print(herzfrequenz)


Deine Aufgabe ist es die folgenden Informationen über die Hunde herausfinden (Abschnitt 8/20).
Dies ist die MUSTERLÖSUNG! Die Coding-Konsole im Ki-Kurs enthält weitere Informationen zu Funktionen, Argumenten oder Erklärungen zum Ergebnis.

In [ ]:
# Das Alter aller Hunde
alter_alle = D[:,0] #hier definierst du eine neue Variable, die Variable alter_alle
print(alter_alle) #der Print Befehl zeigt dir das Ergebnis an!

In [ ]:
# Die ersten drei Merkmale des ersten Hundes
drei_merkmale_hund1 = D[0,0:3]
print(drei_merkmale_hund1)

In [ ]:
# Gesundheitszustand des 14. Hundes
gesundheit_hund14 = D[13,13]
print(gesundheit_hund14)

In [ ]:
 # Alle Merkmale des 14. Hundes
merkmale_hund14 = D[13,0:13]
print(merkmale_hund14)

In [ ]:
# Alle Merkmale aller Hunde
merkmale_alle = D[:,0:13]
print(merkmale_alle)

In [ ]:
#Oder du arbeitest mit einer Definition def()
def ausgaben(D):
    alter = D[:,0]
    drei_merkmale_hund1 = D[0, 0:3]
    gesundheit_hund14 = D[13, -1]
    merkmale_hund14 = D[13, :-1]
    merkmale_alle = D[:, :-1]

    return alter, drei_merkmale_hund1, gesundheit_hund14, merkmale_hund14, merkmale_alle

Nun sollst du dir die Anzahl an gespeicherten Variablen in der Funktion Ausgabe anzeigen lassen 

In [ ]:
len(ausgaben(D))

In [ ]:
#Und nun den numerischen Inhalt der Funktion ausgeben
print(ausgaben(D))

#oder die numerische Ausgabe einer Variable
print(drei_merkmale_hund1)



Deine Aufgabe ist herauszufinden, wie viele der Hunde krank sind (Abschnitt 11/20).
Dies ist die MUSTERLÖSUNG! Die Coding-Konsole im Ki-Kurs enthält weitere Informationen zu Funktionen, Argumenten oder Erklärungen zum Ergebnis.

In [ ]:
#Schreibe eine Funktion, die sowohl den Gesundheitszustand von Hund 1-13 (0=gesund, 1=krank) als auch die Anzahl kranker Hunde definiert
def anzahl_kranker_hunde(D):
    """ Gibt die Anzahl der kranken Hunde aus.
    Args:
        D (numpy.ndarray): Datenmatrix:
            - pro Hund eine Zeile (erster Index)
            - pro Eigenschaft eine Spalte (zweiter Index)
    Returns:
        int: Anzahl kranker Hunde.
    """
   
    anzahl = D[:,-1].sum()
    
    for i in range(0,13):
        print("Der Hund Nummer ", i, "hat den Gesundheitszustand ", D[i, -1])
        
        
    return anzahl

In [ ]:
# Lass dir den Inhalt der Funktion anzahl_kranker_hunde(D) ausgeben
#Diese enthält sowohl den Gesundheitszustand von Hund 1-13 (0=gesund, 1=krank) als auch die Anzahl kranker Hunde (255)
print(anzahl_kranker_hunde(D))


Deine Aufgabe ist, mit Hilfe von der kurzen Funktion np.sum() herauszufinden, wie viele der Hunde krank sind (Abschnitt 12/20).
Dies ist die MUSTERLÖSUNG! Die Coding-Konsole im Ki-Kurs enthält weitere Informationen zu Funktionen, Argumenten oder Erklärungen zum Ergebnis.

In [ ]:
#Nutze anstelle von einer For-Schleife die Funktion np.sum() um die Anzahl kranker Hunde zu berechnen
anzahl = np.sum(D[:,13])
print(anzahl)

Deine Aufgabe ist, dir mit Hilfe von einer if-else-Abfrage in Textform ausgeben zu lassen, ob der Hund gesund (0)oder krank (1) ist oder ein Fehler aufgetreten ist.(Abschnitt 16/20).
Dies ist die MUSTERLÖSUNG! Die Coding-Konsole im Ki-Kurs enthält weitere Informationen zu Funktionen, Argumenten oder Erklärungen zum Ergebnis.

In [ ]:
#das package numpy haben wir bereits weiter oben installiert

def textausgabe(gesundheitszustand):
    """Textausgabe über Gesundheitszustand
    Args:
        gesundheitszustand (int): Gibt den Gesundheitszustand eines Hundes an
        
    Returns:
            (string): Text, der ausgibt:
            Für gesunde Hunde: "Dieser Hund ist gesund."
            Für kranke Hunde: "Dieser Hund ist krank."
            Für Werte ungleich 0 oder 1: "Fehler."
    """
    if gesundheitszustand == 0:
        ausgabe = "Dieser Hund ist gesund."

    elif gesundheitszustand == 1:
        ausgabe = "Dieser Hund ist krank."

    else: 
        ausgabe = "Fehler."

    return ausgabe

In [ ]:
#Lass dir den Gesundheitszustand des ersten Hundes ausgeben
print(textausgabe(D[0,13]))
#Lass dir den Gesundheitszustand des 232. Hundes ausgeben
print(textausgabe(D[231,13]))


Deine Aufgabe ist der Tierärztin, mit Hilfe von einer while-Schleife eine beliebige Anzahl an gesunden oder kranken Hunden aus dem Datensatz ausgeben zu können, die sie untersuchen kann (Abschnitt 17/20). Dies ist die MUSTERLÖSUNG! Die Coding-Konsole im Ki-Kurs enthält weitere Informationen zu Funktionen, Argumenten oder Erklärungen zum Ergebnis.

In [9]:
#das package numpy haben wir bereits weiter oben installiert

def hunde_ausgabe(D, anzahl, gesundheitszustand):
    """ Gibt eine bestimmte Anzahl an Hunden eines gewissen Gesundheitszustandes aus
    Args:
        D (numpy.ndarray): Datenmatrix:
            - pro Hund eine Zeile (erster Index)
            - pro Eigenschaft eine Spalte (zweiter Index)
        anzahl (int): Anzahl an Hunden, die zurück gegeben werden sollen
        gesundheitszustand (float): gewünschter Gesundheitszustand
    Returns:
        list: Liste mit den Indizes der Hunde
    """
    liste_hunde = []
    zähler = 0
    hund = 0

    while zähler < anzahl:
        if D[hund,-1] == gesundheitszustand:
            zähler = zähler + 1
            liste_hunde.append(hund)

        hund = hund + 1

    return liste_hunde




In [10]:
#Hier wird der Tierärztin eine Liste mit 10 kranken Hunden vorgelegt, die sie schnellstmöglich untersuchen sollte
liste = hunde_ausgabe(D,10,1)
print(liste)

[1, 2, 6, 8, 9, 12, 16, 22, 23, 24]


Deine nächste Aufgabe - Das Tierheim möchte wissen, was die durchschnittliche Herzfrequenz (Merkmal 8) aller Hunde ist. Kannst du ihnen dabei helfen (Abschnitt 18/20)? Dies ist die MUSTERLÖSUNG! Die Coding-Konsole im Ki-Kurs enthält weitere Informationen zu Funktionen, Argumenten oder Erklärungen zum Ergebnis.

In [ ]:
#Das Paket numpy haben wir bereits weiter oben importiert
def durchschnittliche_herzfrequenz(D):
    """ Gibt das Mittel der Herzfrequenzen aller Hunde aus.
    Args:
        D (numpy.ndarray): Datenmatrix:
            - pro Hund eine Zeile (erster Index)
            - pro Eigenschaft eine Spalte (zweiter Index)
    Returns:
        float: Arithmetisches Mittel der Herzfrequenzen (mean heart rate = mhr) aller Hunde
    """
    mhr = D[:, 7].mean()
    return mhr

In [ ]:
#Lass dir die durchschnittliche Herzfrequenz aller Hunde ausgeben
print(durchschnittliche_herzfrequenz(D))

In [ ]:
#Wow das ist aber eine lange Zahl .. Kürzen wir diese doch auf 2 Nachkommastellen
round(132.51682867672503, 2)

In der nächsten Aufgabe geht es darum herauszufinden, ob ein Zusammenhang zwischen der Herzfrequenz und dem Gesundheitszustand besteht d.h. haben kranke Hunde (1) im Durchschnitt eine niedrigere oder höhere Herzfrequenz, als Hunde die gesund sind (0) (Abschnitt 19/20)? Dies ist die MUSTERLÖSUNG! Die Coding-Konsole im KI-Kurs enthält weitere Informationen zu Funktionen, Argumenten oder Erklärungen zum Ergebnis.

In [ ]:
#Numpy wurde bereits weiter oben installiert
# Abkürzungserläuterung dhf_krank = filtert alle kranken Hunde heraus
# Abkürzungserläuterung ahr = berechnet die mittlere Herzfrequenz (8.Merkmal) aller kranken Hunde


def durchschnittliche_herzfrequenz_kranker_hunde(D):
    dhf_krank = D[:, -1] == 1 #filtert alle kranken Hunde heraus
    ahr = D[dhf_krank, 7].mean() #berechnet die mittlere Herzfrequenz (8.Merkmal) aller kranken Hunde
    return ahr

def durchschnittliche_herzfrequenz_gesunder_hunde(D):
    dhf_gesund = D[:, -1] == 0 #filtert alle gesunden Hunde heraus
    ahr = D[dhf_gesund, 7].mean() #berechnet die mittlere Herzfrequenz (8.Merkmal) aller kranken Hunde
    return ahr

In der letzten Aufgabe des zweiten Kapitels geht es darum schnellstmöglich herauszufinden, ob ein Hund gesund oder krank ist und zwar direkt bei der Einlieferung der Hunde. Hierfür vergleichen wir die durchschnittliche Herzfrequenz von gesunden Hunden, mit der von kranken Hunden und berechnen die Differenz zwischen den Mittelwerten(Abschnitt 20/20). Dies ist die MUSTERLÖSUNG! Die Coding-Konsole im Ki-Kurs enthält weitere Informationen zu Funktionen, Argumenten oder Erklärungen zum Ergebnis.

In [ ]:
#numpy wurde bereits installiert 

def unterschied_herzfrequenz(D):
    mhr_diff = durchschnittliche_herzfrequenz_gesunder_hunde(D) - durchschnittliche_herzfrequenz_kranker_hunde(D)
    return mhr_diff
    

In [ ]:
#Ausgabe der Aufgaben 19 & 20
#Wie hoch ist die durchschnittliche Herzfrequenz kranker (1) Hunde?
print(durchschnittliche_herzfrequenz_kranker_hunde(D))
#Wie hoch ist die durchschnittliche Herzfrequenz kranker (0) Hunde?
print(durchschnittliche_herzfrequenz_gesunder_hunde(D))
#Wie hoch ist die Differenz in der Herzfrequenz zwischen gesunden und kranken Hunden?
print(unterschied_herzfrequenz(D))

## Du hast es geschafft! 🥳


Wir sind mit dem ersten und zweiten Kapitel durch. Jetzt erst mal etwas Nervennahrung zu sich nehmen und dann kann's schon weiter gehen oder?
